linfa-bayes/src/gaussian_nb.rs

```diff
@@ -0,0 +1,546 @@
+        // If incoming data is empty no updates required
+        if x_new.nrows() == 0 {
+            return (mu_old.to_owned(), var_old.to_owned());
+        }
+
+        let count_new = A::from(x_new.nrows()).unwrap();
+        let mu_new = x_new.mean_axis(Axis(0)).unwrap();
+        let var_new = x_new.var_axis(Axis(0), A::from(0.).unwrap());
+
+        // If previous batch was empty, we send the new mean and variance calculated
+        if count_old == A::from(0.).unwrap() {
+            return (mu_new, var_new);
+        }
+
+        let count_total = count_old + count_new;
+
+        // Combine old and new mean, taking into consideration the number
+        // of observations
+        let mu_new_weighted = &mu_new * A::from(count_new).unwrap();
+        let mu_old_weighted = mu_old * A::from(count_old).unwrap();
+        let mu_weighted = (mu_new_weighted + mu_old_weighted).mapv(|x| x / count_total);
+
+        // Combine old and new variance, taking into consideration the number
+        // of observations. this is achieved by combining the sum of squared
+        // differences
+        let ssd_old = var_old * A::from(count_old).unwrap();
+        let ssd_new = var_new * A::from(count_new).unwrap();
+        let weight = A::from(count_new * count_old).unwrap() / count_total;
+        let ssd_weighted = ssd_old + ssd_new + (mu_old - &mu_new).mapv(|x| weight * x.powi(2));
+        let var_weighted = ssd_weighted.mapv(|x| x / count_total);
+
+        (mu_weighted, var_weighted)
+    }
+
+    // Extract unique elements of the array in sorted order
+    fn unique(y: &ArrayView1<A>) -> Array1<A> {
+        // We are identifying unique classes in y,
+        // ndarray doesn't provide methods for extracting unique elements,
+        // So we are converting it to a Vec
+        let mut unique_classes = y.to_vec();
+        unique_classes.sort_by(|x, y| x.partial_cmp(y).unwrap());
+        unique_classes.dedup();
+
+        Array1::from(unique_classes)
+    }
+
+    // Returns a subset of x corresponding to the class specified by `ycondition`
+    fn filter(x: &ArrayView2<A>, y: &ArrayView1<A>, ycondition: A) -> Result<Array2<A>> {
+        // We identify the row numbers corresponding to the class we are interested in
+        let index: Vec<_> = y
+            .indexed_iter()
+            .filter_map(|(i, y)| {
+                if (ycondition - *y).abs() < A::from(1e-6).unwrap() {
+                    return Some(i);
+                }
+                None
+            })
+            .collect();
+
+        // We subset x to only records corresponding to the class represented in `ycondition`
+        let mut xsubset = Array2::zeros((index.len(), x.ncols()));
+        index
+            .iter()
+            .enumerate()
+            .for_each(|(i, &r)| xsubset.row_mut(i).assign(&x.slice(s![r, ..])));
+
+        Ok(xsubset)
+    }
+}
+
+/// Fitted GaussianNB for predicting classes
+#[derive(Debug)]
+pub struct FittedGaussianNb<A> {
```

Comment:
this should just be `GaussianNb`, which was previously the hyperparmeter struct

Reply:
Changed